stackrox/stackrox
# ROX-19048: Integrate Scanner V4 into secured-cluster-services Chart

## Description

This PR adds Scanner V4 support to the secured-cluster-services Helm chart.
Since secured clusters can run in "local scanner" mode, which depends on scanner certificates being issued by central, this PR also extends the certificate issuing code within central.

## Checklist

Feature is not public yet.

## Testing Performed

### Here I tell how I validated my change

Manual testing:

1. Preparations:
a. Connect to cluster, e.g. Infra GKE with instance type `e2-standard-8`
b. Make sure `$QUAY_USERNAME` and `$QUAY_TOKEN` contain valid quay.io credentials.
c. Set `export MAIN_IMAGE_TAG=...`.
d. Prepare the charts:
```
$ make cli
$ roxctl helm output central-services --output-dir ./stackrox-central-services-chart --remove --debug
$ roxctl helm output secured-cluster-services --output-dir ./stackrox-secured-cluster-services-chart --remove --debug
```
2. Install central-services:
a. I use the following chart configuration stored as `values-scannerv4.yaml`:
```
central:
  resources:
    requests:
      cpu: 400m
      memory: 1Gi
    limits:
      cpu: 400m
      memory: 1Gi
  image:
    tag: "$MAIN_IMAGE_TAG"
  persistence:
    none: true
  db:
    resources:
      requests:
        cpu: 400m
        memory: 1Gi
      limits:
        cpu: 400m 
        memory: 1Gi
    image:
      tag: "$MAIN_IMAGE_TAG"
  
  telemetry:
    enabled: false 

scanner:
  disable: false #true
  
scannerV4:
  disable: false 
  indexer:
    image:
      tag: "$MAIN_IMAGE_TAG"
  matcher:
    image:
      tag: "$MAIN_IMAGE_TAG"
  db:
    image:
      tag: "$MAIN_IMAGE_TAG"
  
imagePullSecrets:
  username: "$QUAY_USER"
  password: "$QUAY_TOKEN"

allowNonstandardNamespace: true
```
b. Set namespace: `export NS=stackrox`
c. run helm install:
```
helm install -n "$NS" --create-namespace --set central.exposure.loadBalancer.enabled=true --values <(envsubst < values-scannerv4.yaml) stackrox-central-services ./stackrox-central-services-chart
```
d. Set password: `export ROX_ADMIN_PASSWORD=<as produced by helm install>`
e. Obtain central endpoint:
```
export CENTRAL_ENDPOINT="https://$(kubectl -n stackrox get service central-loadbalancer -o jsonpath='{.status.loadBalancer.ingress[0].ip}'):443"
```
3. Install secured-cluster-services
a. I use the following chart configuration stored as `values-scannerv4-sc.yaml`:
```
clusterName: sc0
centralEndpoint: "$CENTRAL_ENDPOINT" 

image:
  main:
    tag: "$MAIN_IMAGE_TAG"
  collector:
    tag: "$MAIN_IMAGE_TAG"
  scannerV4:
    tag: "$MAIN_IMAGE_TAG"
  scannerV4DB:
    tag: "$MAIN_IMAGE_TAG"

sensor:
  resources:
    requests:
      cpu: 500m
      memory: 1Gi
    limits:
      cpu: 1000m
      memory: 2Gi

scanner:
  disable: false

scannerV4:
  disable: false 

imagePullSecrets:
  username: "$QUAY_USER"
  password: "$QUAY_TOKEN"

allowNonstandardNamespace: true
```
b. Set namespace: `export NS=stackrox-sc`
c. retrieve init bundle:
```
roxctl -e "$CENTRAL_ENDPOINT" --insecure-skip-tls-verify central init-bundles generate sc0 --output init-bundle-sc0.yaml
```
d. run helm install:
```
helm install -n "$NS" --create-namespace --values <(envsubst < values-scannerv4-sc.yaml) --set centralEndpoint="$CENTRAL_ENDPOINT" -f init-bundle-sc0.yaml stackrox-secured-cluster-services ./stackrox-secured-cluster-services-chart
```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.
